crates/goose/src/session/storage.rs

```diff
@@ -1356,6 +1542,100 @@ pub async fn update_metadata(session_file: &Path, metadata: &SessionMetadata) ->
     save_messages_with_metadata(&secure_path, metadata, &messages)
 }
 
+/// Persist messages in the background without blocking
+pub fn persist_messages_background(
+    session_file: &Path,
+    messages: &Conversation,
+    provider: Option<Arc<dyn Provider>>,
+    working_dir: Option<PathBuf>,
+) -> Result<()> {
+    persist_messages_with_schedule_id_background(
+        session_file,
+        messages,
+        provider,
+        None,
+        working_dir,
+    )
+}
+
+/// Persist messages with schedule ID in the background without blocking
+pub fn persist_messages_with_schedule_id_background(
+    session_file: &Path,
+    messages: &Conversation,
+    provider: Option<Arc<dyn Provider>>,
+    schedule_id: Option<String>,
+    working_dir: Option<PathBuf>,
+) -> Result<()> {
+    // Validate the session file path for security
+    let secure_path = get_path(Identifier::Path(session_file.to_path_buf()))?;
+
+    // Security check: message count limit
+    if messages.len() > MAX_MESSAGE_COUNT {
+        tracing::warn!("Message count exceeds limit: {}", messages.len());
+        return Err(anyhow::anyhow!("Too many messages"));
+    }
+
+    // Count user messages
+    let user_message_count = messages
+        .iter()
+        .filter(|m| m.role == rmcp::model::Role::User && !m.as_concat_text().trim().is_empty())
+        .count();
+
+    // Create metadata immediately (synchronous)
+    let mut metadata = if secure_path.exists() {
+        read_metadata(&secure_path)?
+    } else {
+        // Create new metadata with the provided working_dir or fall back to home
+        let work_dir = working_dir.clone().unwrap_or_else(get_home_dir);
+        SessionMetadata::new(work_dir)
+    };
+
+    // Update the working_dir if provided (even for existing files)
+    if let Some(ref work_dir) = working_dir {
+        metadata.working_dir = work_dir.clone();
+    }
+
+    // Update the schedule_id if provided
+    if schedule_id.is_some() {
+        metadata.schedule_id = schedule_id.clone();
+    }
+
+    // Create the task for background saving/naming
+    let task = if let Some(provider) = provider.as_ref() {
+        if user_message_count < 4 {
+            // For early messages that need session naming, use the GenerateName task
+            PersistenceTask::GenerateName {
```

Comment:
nit: I first thought creating only a GenerateName task would mean we don't save, maybe better named as `GenerateNameAndSave` or something like that